packages/core/src/storages/request_queue_v2.ts

```diff
@@ -154,13 +176,85 @@ export class RequestQueue extends RequestProvider {
         //    will not put the request again to queueHeadDict.
         if (request.handledAt) {
             this.log.debug('Request fetched from the beginning of queue was already handled', { nextRequestId });
-            this.recentlyHandledRequestsCache.add(nextRequestId, true);
             return null;
         }
 
         return request;
     }
 
+    /**
+     * @inheritDoc
+     */
+    override async isFinished(): Promise<boolean> {
+        if (this.inProgressRequestBatches.length > 0) {
+            return false;
+        }
+
+        if (this.queueHeadIds.length() > 0) {
+            return false;
+        }
+
+        await this.ensureHeadIsNonEmpty();
+
+        if (this.queueHeadIds.length() > 0) {
+            return false;
+        }
+
+        if (this.queueHasLockedRequests !== undefined) {
+            // The `% 25` was absolutely arbitrarily picked. It's just to not spam the logs too much.
+            if (this.queueHasLockedRequests && ++this.isFinishedCalledWhileHeadWasNotEmpty % 25 === 0) {
+                this.log.info('The queue still contains requests locked by another client');
+            }
+
+            return !this.queueHasLockedRequests;
+        }
+
+        // The following is a legacy algorithm for checking if the queue is finished
```

Comment:
what does this mean exactly? do we still need it, or is it a failsafe?

Reply:
It's for storage clients that don't provide the `queueHasLockedRequests` property. Currently, this includes the memory storage and possibly other storage client implementations with which we don't want to break compatibility.